lib/internal/modules/esm/default_resolve.js

```diff
@@ -86,26 +86,16 @@ function resolve(specifier, parentURL) {
   let format = extMap[ext];
 
   if (isMain && typeFlag) {
-    // Conflict between explicit extension (.mjs, .cjs) and --entry-type
-    if (ext === '.cjs' && typeFlag === 'module' ||
-        ext === '.mjs' && typeFlag === 'commonjs') {
-      throw new ERR_ENTRY_TYPE_MISMATCH(
-        fileURLToPath(url), ext, typeFlag, 'extension');
-    }
-
-    // Conflict between package scope type and --entry-type
-    if (ext === '.js') {
-      if (type === TYPE_MODULE && typeFlag === 'commonjs' ||
-          type === TYPE_COMMONJS && typeFlag === 'module') {
-        throw new ERR_ENTRY_TYPE_MISMATCH(
-          fileURLToPath(url), ext, typeFlag, 'scope');
-      }
-    }
+    // This is the initial entry point to the program, and --input-type has
+    // been passed as an option; but --input-type can only be used with
+    // --eval, --print or STDIN string input. It is not allowed with file
+    // input, to avoid user confusion over how expansive the effect of the
+    // flag should be (i.e. entry point only, package scope surrounding the
+    // entry point, etc.).
+    throw new ERR_INPUT_TYPE_NOT_ALLOWED();
```

Comment:
So instead of catching the error here, it should be caught much much earlier in the Node bootstrap itself.